src/changeset.rs

```diff
@@ -0,0 +1,195 @@
+        {
+            Some(_) => panic!("tried to add two of the same type of change"),
+            None => self.changes.push(Change::new(change, validations)),
+        }
+        self
+    }
+    pub fn validate(&mut self) -> Result<&mut Self, Vec<&'static str>> {
+        let mut errors_acc = vec![];
+
+        let errors = self.changes.iter().fold(
+            &mut errors_acc,
+            |error_list, Change(attribute, validations)| {
+                let mut errors_for_element: Vec<&str> = validations
+                    .iter()
+                    .filter_map(|validation| validation(attribute).err())
+                    .collect();
+                error_list.append(&mut errors_for_element);
+                error_list
+            },
+        );
+        if errors.len() > 0 {
+            Err(errors_acc)
+        } else {
+            self.valid = true;
+            Ok(self)
+        }
+    }
+    // pub async fn insert(self, pool: &PgPool) -> Result<D, Error> {
+    //     pool.fetch_one(self.build_query().as_str())
+    //         .await
+    //         .map(|row| Data::from_row(&row).unwrap())
+    // }
+    // fn build_query(&self) -> String {
+    //     Query::insert()
+    //         .into_table(Data::table_ref())
+    //         .columns(self.changes.clone())
+    //         .values_panic(self.changes.clone().iter().map(|x| x.value()))
+    //         .to_string(PostgresQueryBuilder)
+    // }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::entities::rfc::{RFCAttrs, Status, RFC};
+    //TODO create a bogus enum and stuff here
+    fn setup() -> Changeset<RFCAttrs, RFC> {
+        let cs: Changeset<RFCAttrs, RFC> = Changeset::new(None);
+        cs
+    }
+    mod add_change_tests {
+        use super::super::*;
+        use crate::entities::rfc::{RFCAttrs, Status, RFC};
+
+        #[test]
+        #[should_panic]
+        fn will_not_accept_duplicate_change_types() {
+            let mut cs = super::setup();
+            cs.add_change(RFCAttrs::Status(Status::Active), None);
+            cs.add_change(RFCAttrs::Status(Status::Active), None);
+        }
+
+        #[test]
+        fn will_accept_distinct_change_types() {
+            let mut cs = super::setup();
+            cs.add_change(RFCAttrs::Status(Status::Active), None);
+            cs.add_change(RFCAttrs::Proposal("Whatever".to_string()), None);
+            let changes: Vec<RFCAttrs> = cs.changes.iter().map(|change| change.0.clone()).collect();
+
+            assert_eq!(
+                vec![
+                    RFCAttrs::Status(Status::Active),
+                    RFCAttrs::Proposal("Whatever".to_string())
+                ],
+                changes
+            )
+        }
+    }
+    mod validate_tests {
+        use std::panic::panic_any;
+
+        use super::super::*;
+        use crate::entities::rfc::{RFCAttrs, Status, RFC};
+        // #[test]
+        fn fails_if_required_attrs_not_present() {
+            // Desired behaviour
```

Comment:
This comment is kind of before I implemented validation logic. Now I think the big challenge is making certain fields required. That could be challenging since if you have backing data that has those fields you'd probably want to consider that requirement filled, but currently there's no like between the attribute enum and the struct it's meant to create/represent.